### PR TITLE
Adds IPC charger compatible with power cells

### PR DESCRIPTION
**Pull request by Yata9ar4su**
## What Does This PR Do

Adding new type of IPC chargers: Compatible with cells.

## Why It's Good For The Game

IPC will no longer be dependent on APC so they can travel without having an issue with charge.
## Images of changes
image
image
image

## Testing

Tested, werks
## Changelog
:cl:
add: Added charger for IPCs that compatible with cells
/:cl:

### CODE REVIEW

File: code/modules/research/designs/mechfabricator_designs.dm [hunk 1]
```diff
@@ -1197,14 +1197,23 @@
...
+	name = "IPC Charger for APCs"
+	id = "ipc_charger"
 	build_type = MECHFAB
 	build_path = /obj/item/organ/internal/cyberimp/arm/power_cord
 	materials = list(MAT_METAL=2000, MAT_GLASS=1000)
 	construction_time = 200
 	category = list("Misc")
 
+/datum/design/ipc_charger_cell
+	name = "IPC Charger for Cells"
```
Comment (1080pCat) [inline code/modules/research/designs/mechfabricator_designs.dm:1209]:
```suggestion
	name = "Power-cell compatible IPC charger"
```

File: code/modules/research/designs/mechfabricator_designs.dm [hunk 2]
```diff
@@ -1197,14 +1197,23 @@
 	category = list("Misc")
 
 /datum/design/ipc_charger
-	name = "IPC Charger"
-	id = "ipc_cahrger"
+	name = "IPC Charger for APCs"
```
Comment (1080pCat) [inline code/modules/research/designs/mechfabricator_designs.dm:1200]:
```suggestion
	name = "APC compatible IPC charger"
```

**Review by 1080pCat — COMMENTED** [review]
Just a bit of a nit pick related to the mech fab design names

File: code/modules/surgery/organs/augments_arms.dm [hunk 3]
```diff
@@ -604,3 +604,75 @@
 	if(emp_proof)
 		return
 	muscle_implant.emp_act(severity, owner)
+
+/obj/item/organ/internal/cyberimp/arm/power_cord_cell
+	name = "Cell-compatible power adapter implant"
+	desc = "This is a recently developed implant for installation inside IPCs in order to allow them to easily collect energy from cells"
+	origin_tech = "materials=3;engineering=4;biotech=3"
```
Comment (1080pCat) [inline code/modules/surgery/organs/augments_arms.dm:611]: One more thing. you've given it quite a high engineering origin but there's no requirement for tech on the mechfab side of things, i'd suggest making it atleast the same as the apc one.

**Review by 1080pCat — CHANGES_REQUESTED** [review]

**Comment by Yata9ar4su** [comment]
> I feel like this'd do well as a option you can pick as preference, like how on vox/plasmamen you can swap between large tank and pocket tank. as well as being in the mech fab

Everyone will just pick Cell one. But i feel like it should be on a level with like nutriment power or smth. Oh wait, thats an idea for tesla and plasma generators for IPC's!

File: code/modules/surgery/organs/augments_arms.dm (hunk 3 above)
Comment (Yata9ar4su) [reply code/modules/surgery/organs/augments_arms.dm:611]: > One more thing. you've given it quite a high engineering origin but there's no requirement for tech on the mechfab side of things, i'd suggest making it atleast the same as the apc one.

image
Strange. I have different techs

Comment (Yata9ar4su) [reply code/modules/surgery/organs/augments_arms.dm]: Dun. Now original  materials=3;biotech=2;powerstorage=3

**Comment by Yata9ar4su** [comment]
Omg my code  working…

File: code/modules/surgery/organs/augments_arms.dm [hunk 4]
```diff
@@ -604,3 +604,75 @@
...
+
+/obj/item/power_cord_cell/proc/cell_powerdraw_loop(obj/item/stock_parts/cell/C, mob/living/carbon/human/H)
+	H.visible_message("<span class='notice'>[H] inserts a power connector into \the [C].</span>", "<span class='notice'>You begin to draw power from \the [C].</span>")
+	drawing_power = TRUE
+	while(do_after(H, 10, target = C))
+		if(loc != H)
+			to_chat(H, "<span class='warning'>You must keep your connector out while charging!</span>")
+			break
+		if(C.charge == 0)
+			to_chat(H, "<span class='warning'>\The [C] has no more charge.</span>")
```
Comment (DGamerL) [inline code/modules/surgery/organs/augments_arms.dm:663]:
```suggestion
			to_chat(H, "<span class='warning'>[C] has no more charge.</span>")
```
This shouldn't need `\The` here I'm pretty sure

File: code/modules/surgery/organs/augments_arms.dm [hunk 5]
```diff
@@ -604,3 +604,75 @@
...
+			break
+		if(C.charge >= 500)
+			H.adjust_nutrition(10)
+			C.charge -= 500
+			to_chat(H, "<span class='notice'>You siphon off some of the stored charge for your own use.</span>")
+		else
+			H.adjust_nutrition(C.charge * 20)
+			C.charge = 0
+			to_chat(H, "<span class='notice'>You siphon off the last of \the [C]'s charge.</span>")
```
Comment (DGamerL) [inline code/modules/surgery/organs/augments_arms.dm:672]:
```suggestion
			to_chat(H, "<span class='notice'>You siphon off the last of [C]'s charge.</span>")
```
Same here

File: code/modules/surgery/organs/augments_arms.dm [hunk 6]
```diff
@@ -604,3 +604,75 @@
...
+		return
+	damage = 1
+	crit_fail = TRUE
+
+/obj/item/organ/internal/cyberimp/arm/power_cord_cell/surgeryize()
+	if(crit_fail && owner)
+		to_chat(owner, "<span class='notice'>Your [src] feels functional again.</span>")
+	crit_fail = FALSE
+
+/obj/item/power_cord_cell
```
Comment (Henri215) [inline code/modules/surgery/organs/augments_arms.dm:627]: Could probably rename that path name so its not the same as the implant one.

File: code/modules/surgery/organs/augments_arms.dm [hunk 7]
```diff
@@ -604,3 +604,75 @@
...
+	name = "power cable"
+	desc = "Insert into a cell to draw power from it."
+	icon = 'icons/obj/power.dmi'
+	icon_state = "wire1"
+	flags = NOBLUDGEON
+	var/drawing_power = FALSE
```
Comment (Henri215) [inline code/modules/surgery/organs/augments_arms.dm:633]:
```suggestion
	/// Are we currently drawing power from a cell?
	var/drawing_power = FALSE
```

File: code/modules/surgery/organs/augments_arms.dm [hunk 8]
```diff
@@ -604,3 +604,75 @@
...
+				to_chat(user, "<span class='warning'>You are already fully charged!</span>")
+			else
+				INVOKE_ASYNC(src, PROC_REF(cell_powerdraw_loop), C, H)
+		else
+			to_chat(user, "<span class='warning'>There is no charge to draw from that cell.</span>")
+	else
+		to_chat(user, "<span class='warning'>You lack a cell in which to store charge!</span>")
+
+/obj/item/power_cord_cell/proc/cell_powerdraw_loop(obj/item/stock_parts/cell/C, mob/living/carbon/human/H)
+	H.visible_message("<span class='notice'>[H] inserts a power connector into \the [C].</span>", "<span class='notice'>You begin to draw power from \the [C].</span>")
```
Comment (Henri215) [inline code/modules/surgery/organs/augments_arms.dm:656]:
```suggestion
	H.visible_message("<span class='notice'>[H] inserts a power connector into [C].</span>", "<span class='notice'>You begin to draw power from [C].</span>")
```

File: code/modules/surgery/organs/augments_arms.dm [hunk 9]
```diff
@@ -604,3 +604,75 @@
...
+			to_chat(H, "<span class='notice'>You siphon off some of the stored charge for your own use.</span>")
+		else
+			H.adjust_nutrition(C.charge * 20)
+			C.charge = 0
+			to_chat(H, "<span class='notice'>You siphon off the last of \the [C]'s charge.</span>")
+			break
+		if(H.nutrition > NUTRITION_LEVEL_WELL_FED)
+			to_chat(H, "<span class='notice'>You are now fully charged.</span>")
+			break
+	H.visible_message("<span class='notice'>[H] unplugs from \the [C].</span>", "<span class='notice'>You unplug from \the [C].</span>")
```
Comment (Henri215) [inline code/modules/surgery/organs/augments_arms.dm:677]:
```suggestion
	H.visible_message("<span class='notice'>[H] unplugs from [C].</span>", "<span class='notice'>You unplug from [C].</span>")
```

File: code/modules/surgery/organs/augments_arms.dm (hunk 6 above)
Comment (Yata9ar4su) [reply code/modules/surgery/organs/augments_arms.dm]: > Could probably rename that path name so its not the same as the implant one.

dun

File: code/modules/surgery/organs/augments_arms.dm [hunk 10]
```diff
@@ -604,3 +604,76 @@
...
+	if(H.get_int_organ(/obj/item/organ/internal/cell))
+		if(C && C.charge > 0)
+			if(H.nutrition >= NUTRITION_LEVEL_WELL_FED)
+				to_chat(user, "<span class='warning'>You are already fully charged!</span>")
+			else
+				INVOKE_ASYNC(src, PROC_REF(cell_powerdraw_loop), C, H)
+		else
+			to_chat(user, "<span class='warning'>There is no charge to draw from that cell.</span>")
+	else
+		to_chat(user, "<span class='warning'>You lack a cell in which to store charge!</span>")
```
Comment (DGamerL) [inline code/modules/surgery/organs/augments_arms.dm:654]: Rewrite this using guard clauses please

File: code/modules/surgery/organs/augments_arms.dm [hunk 11]
```diff
@@ -604,3 +604,76 @@
...
+	drawing_power = TRUE
+	while(do_after(H, 10, target = C))
+		if(loc != H)
+			to_chat(H, "<span class='warning'>You must keep your connector out while charging!</span>")
+			break
+		if(C.charge == 0)
+			to_chat(H, "<span class='warning'>[C] has no more charge.</span>")
+			break
+		if(C.charge >= 500)
+			H.adjust_nutrition(10)
```
Comment (DGamerL) [inline code/modules/surgery/organs/augments_arms.dm:667]: See Qwerty's comment in the discord

**Review by 1080pCat — APPROVED** [review]
Code looks good, just address DGL's review.

File: code/modules/surgery/organs/augments_arms.dm [hunk 12]
```diff
@@ -604,3 +604,76 @@
...
+
+/obj/item/organ/internal/cyberimp/arm/power_cord_cell
+	name = "cell-compatible power adapter implant"
+	desc = "This is a recently developed implant for IPCs in order to allow them to easily collect energy from cells."
+	origin_tech = "materials=3;biotech=2;powerstorage=3"
+	contents = newlist(/obj/item/cell_power_cord)
+	requires_robotic_bodypart = TRUE
+
+/obj/item/organ/internal/cyberimp/arm/power_cord_cell/emp_act(severity)
+// Allowing to repair implant
```
Comment (Sirryan2002) [inline code/modules/surgery/organs/augments_arms.dm:616]:
```suggestion
	// Allowing to repair implant
```
indentation

**Review by Sirryan2002 — COMMENTED** [review]
address DGamerL's review

File: code/modules/surgery/organs/augments_arms.dm (hunk 10 above)
Comment (Yata9ar4su) [reply code/modules/surgery/organs/augments_arms.dm:654]: Dun

File: code/modules/research/designs/mechfabricator_designs.dm [hunk 13]
```diff
@@ -1197,14 +1197,23 @@
...
 	construction_time = 200
 	category = list("Misc")
 
+/datum/design/ipc_charger_cell
+	name = "Power-Cell Compatible IPC Charger"
+	id = "ipc_charger_cell"
+	build_type = MECHFAB
+	build_path = /obj/item/organ/internal/cyberimp/arm/power_cord_cell
+	materials = list(MAT_METAL = 2000, MAT_GLASS = 1000)
+	construction_time = 200
```
Comment (Henri215) [inline code/modules/research/designs/mechfabricator_designs.dm:1214]:
```suggestion
	materials = list(MAT_METAL = 2000, MAT_GLASS = 1000)
	construction_time = 20 SECONDS
	category = list("Misc")

/datum/design/ipc_charger_cell
	name = "Power-Cell Compatible IPC Charger"
	id = "ipc_charger_cell"
	build_type = MECHFAB
	build_path = /obj/item/organ/internal/cyberimp/arm/power_cord_cell
	materials = list(MAT_METAL = 2000, MAT_GLASS = 1000)
	construction_time = 20 SECONDS
```

File: code/modules/surgery/organs/augments_arms.dm (hunk 10 above)
Comment (Henri215) [reply code/modules/surgery/organs/augments_arms.dm:654]: Dun?

File: code/modules/surgery/organs/augments_arms.dm [hunk 14]
```diff
@@ -604,3 +604,76 @@
 	if(emp_proof)
 		return
 	muscle_implant.emp_act(severity, owner)
+
+/obj/item/organ/internal/cyberimp/arm/power_cord_cell
+	name = "Cell-compatible power adapter implant"
```
Comment (Henri215) [inline code/modules/surgery/organs/augments_arms.dm:609]: Made the first letter capitalized again?

**Review by CinnamonSnowball — COMMENTED** [review]
There's a lot of unfortunate duplicated code due to the two implants being separate items rather than one of them being a subtype of the other with some values switched, besides that a couple things that seem a bit odd

File: code/modules/mob/living/carbon/human/species/machine.dm [hunk 15]
```diff
@@ -51,7 +51,7 @@
 		"brain" = /obj/item/organ/internal/brain/mmi_holder/posibrain,
 		"cell" = /obj/item/organ/internal/cell,
 		"eyes" = /obj/item/organ/internal/eyes/optical_sensor, //Default darksight of 2.
-		"charger" = /obj/item/organ/internal/cyberimp/arm/power_cord
+		"charger" = /obj/item/organ/internal/cyberimp/arm/power_cord_apc
```
[the remark this answers is not in the document]
Comment (Yata9ar4su) [reply code/modules/mob/living/carbon/human/species/machine.dm]: Yep, no problembs here

File: code/modules/antagonists/traitor/contractor/datums/syndicate_contract.dm [hunk 16]
```diff
@@ -352,7 +352,7 @@
 		if(isgrey(H) && istype(I, /obj/item/organ/internal/cyberimp/brain/speech_translator))
 			continue
 		// IPCs keep this implant, free of charge!
-		if(ismachineperson(H) && istype(I, /obj/item/organ/internal/cyberimp/arm/power_cord))
+		if(ismachineperson(H) && istype(I, /obj/item/organ/internal/cyberimp/arm/power_cord_apc))
```
[the remark this answers is not in the document]
Comment (CinnamonSnowball) [reply code/modules/antagonists/traitor/contractor/datums/syndicate_contract.dm]: ..is this intentional? I don't see any particular reason only one of the implants should be allowed

File: code/modules/antagonists/traitor/contractor/datums/syndicate_contract.dm [hunk 17]
```diff
@@ -352,7 +352,7 @@
 		if(isgrey(H) && istype(I, /obj/item/organ/internal/cyberimp/brain/speech_translator))
 			continue
 		// IPCs keep this implant, free of charge!
-		if(ismachineperson(H) && istype(I, /obj/item/organ/internal/cyberimp/arm/power_cord))
+		if(ismachineperson(H) && istype(I, /obj/item/organ/internal/cyberimp/arm/power_cord/apc))
```
Comment (DGamerL) [inline code/modules/antagonists/traitor/contractor/datums/syndicate_contract.dm:355]:
```suggestion
		if(ismachineperson(H) && istype(I, /obj/item/organ/internal/cyberimp/arm/power_cord))
```
I think this should make it universal

File: code/modules/mob/living/carbon/human/species/machine.dm [hunk 18]
```diff
@@ -232,3 +232,14 @@
...
+
+/datum/species/machine/after_equip_job(datum/job/J, mob/living/carbon/human/H)
+	var/charger_pref = H.client.prefs.active_character.speciesprefs
+	var/obj/item/organ/internal/cyberimp/arm/power_cord/apc/apc_pw = new
+	var/obj/item/organ/internal/cyberimp/arm/power_cord/cell/cell_pw = new
+	if(charger_pref)
+		charger_pref = new /obj/item/organ/internal/cyberimp/arm/power_cord/apc(H)
+		cell_pw.insert(H)
+	else
+		charger_pref = new /obj/item/organ/internal/cyberimp/arm/power_cord/cell(H)
```
Comment (DGamerL) [inline code/modules/mob/living/carbon/human/species/machine.dm:244]: You are still making 3 implants here, where you only need one. Please look at my suggested code from a while ago and try to incorporate that, I'd be glad to help you out again.

File: code/modules/mob/living/carbon/human/species/machine.dm [hunk 19]
```diff
@@ -232,3 +232,14 @@
 								)
 		var/error_message = pick(error_messages)
 		to_chat(H, "<span class='boldwarning'>[error_message]</span>")
+
+/datum/species/machine/after_equip_job(datum/job/J, mob/living/carbon/human/H)
+	var/charger_pref = H.client.prefs.active_character.speciesprefs
+	var/obj/item/organ/internal/cyberimp/arm/power_cord/apc/apc_pw = new
+	var/obj/item/organ/internal/cyberimp/arm/power_cord/cell/cell_pw = new
+	if(charger_pref)
```
Comment (DGamerL) [inline code/modules/mob/living/carbon/human/species/machine.dm:240]: Won't this always give you an APC powercord? Due to strings always being `TRUE`?

**Review by DGamerL — CHANGES_REQUESTED** [review]

File: code/modules/surgery/organs/augments_arms.dm [hunk 20]
```diff
@@ -349,36 +349,36 @@
...
 	crit_fail = TRUE
 
-/obj/item/organ/internal/cyberimp/arm/power_cord_apc/surgeryize()
+/obj/item/organ/internal/cyberimp/arm/power_cord/apc/surgeryize()
 	if(crit_fail && owner)
 		to_chat(owner, "<span class='notice'>Your [src] feels functional again.</span>")
 	crit_fail = FALSE
 
 
-/obj/item/apc_powercord
```
Comment (CinnamonSnowball) [inline code/modules/surgery/organs/augments_arms.dm:373]: You made these two `/obj/item/powercord/apc` and `/obj/item/powercord/cell` subtypes of a ghost blank `/obj/item/powercord` type while leaving duplicated code on both, which isn't great

You should extract the common functionality between the two subtypes into this abstract parent type, and only leave the distinct differences on the subtypes - if you need some pointers on what this means and how to do it, feel free to ask on discord

File: code/modules/mob/living/carbon/human/species/machine.dm [hunk 21]
```diff
@@ -232,3 +232,14 @@
...
+		charger_pref = new /obj/item/organ/internal/cyberimp/arm/power_cord/apc(H)
+		cell_pw.insert(H)
+	else
+		charger_pref = new /obj/item/organ/internal/cyberimp/arm/power_cord/cell(H)
+		apc_pw.insert(H)
```
Comment (DGamerL) [inline code/modules/mob/living/carbon/human/species/machine.dm:245]:
```suggestion
/datum/species/machine/after_equip_job(datum/job/J, mob/living/carbon/human/H)
	var/charger_pref = H.client.prefs.active_character.speciesprefs
	if(charger_pref)
		var/obj/item/organ/internal/cyberimp/arm/power_cord/cell/cell_pw = new(H)
		cell_pw.insert(H)
	else
		var/obj/item/organ/internal/cyberimp/arm/power_cord/apc/apc_pw = new(H)
		apc_pw.insert(H)
```
Try this

File: code/modules/mob/living/carbon/human/species/machine.dm [hunk 22]
```diff
@@ -232,3 +232,11 @@
 								)
 		var/error_message = pick(error_messages)
 		to_chat(H, "<span class='boldwarning'>[error_message]</span>")
+/datum/species/machine/after_equip_job(datum/job/J, mob/living/carbon/human/H)
```
Comment (lewcc) [inline code/modules/mob/living/carbon/human/species/machine.dm:235]:
```suggestion

/datum/species/machine/after_equip_job(datum/job/J, mob/living/carbon/human/H)
```

**Review by lewcc — CHANGES_REQUESTED** [review]
Going to second Snowball's review -- You've duplicated a lot of code under a new type, and you can make your life easier here by taking advantage of inheritance.

I'd recommend keeping the base powercord intact and just adding `/obj/item/powercord/cell` as a subtype. You can then only modify the behavior that changes in this new type.

If you want some advice, here's what I would do:
- Rework the base powerdraw loop to generically pull from a cell. You can just pass the APC's cell in as an argument for the loop in the base call.
- Add some external helper function called in afterattack() that checks if the thing we're attacking is the correct thing, either a cell or an APC

**Review by S34NW — COMMENTED** [review]
No movement in a month, please address reviews in a week or this gets closed due to inactivity.

**Comment by S34NW** [comment]
No changes in the requested timeframe, closing due to being stale.
Please feel free to make a new PR with this once the requested changes have been implemented, you can also reach out to our coding community for help at https://discord.gg/paradisess13

**Pull request closed without merging**